controllers/perses_controller_test.go

```diff
@@ -0,0 +1,202 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	persesv1alpha1 "github.com/perses/perses-operator/api/v1alpha1"
+	common "github.com/perses/perses-operator/internal/perses/common"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ = Describe("Perses controller", func() {
```

Comment:
ah I forgot to comment about this. I'm thinking the same, it's more a php/javascript approach I believe.

The "Golang way" is more about creating a list of cases and then looping other it to run the tested function.
Like that for example: https://github.com/perses/perses/blob/main/pkg/model/api/v1/metadata_test.go#L24-L47

Or for more complicated use case that requires more struct, then you could have something like that:
https://github.com/perses/perses/blob/main/internal/cli/test/test.go

Which is used like that: https://github.com/perses/perses/blob/main/internal/cli/cmd/apply/apply_test.go

Reply:
Thnx for your feedback, I followed the docs on [testing operators](https://sdk.operatorframework.io/docs/building-operators/golang/testing/). In which test are run using `envtest` controller runtime so the eventual state of the operator can be asserted, I believe the goal here is not to test expected outputs of a function as the reconcilers creates resources in a cluster as a side effect. Are there other approaches or suggestions @simonpasquier?

Reply:
I have no strong opinion on the topic. In fact all the operators I've been working on have developed their own home-made framework (based on the Go testing library) to run end-to-end tests. In my current experiences, the end-to-end tests require a running Kube cluster (typically Kind), they setup all required resources (e.g. CRDs, service accounts, roles, bindings, ...) then they run the operator binary and exercise various scenarios.

So I've got no direct experience with the [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest) framework and I understand that it can feel a bit daunting at first but at the same time, it may simplify the setup and teardown of the test environment.